src/librustc/hir/check_attr.rs

```diff
@@ -93,9 +93,10 @@ struct CheckAttrVisitor<'tcx> {
 impl CheckAttrVisitor<'tcx> {
     /// Checks any attribute.
     fn check_attributes(&self, item: &hir::Item, target: Target) {
-        if target == Target::Fn || target == Target::Const {
+        let target_feature = item.attrs.iter().find(|a| a.check_name(sym::target_feature));
+        if target == Target::Fn || (target == Target::Const && target_feature.is_none()) {
             self.tcx.codegen_fn_attrs(self.tcx.hir().local_def_id(item.hir_id));
```

Comment:
Why do we call `codegen_fn_attrs` at all for consts? Most of the codegen attributes don't make any sense for constants. If any are allowed on consts, we should probably have separate handling for them. Only `ffi_returns_twice` seems to throw an error for non-function items.

[This PR added the `target == Target::Const`](https://github.com/rust-lang/rust/pull/51966/files#diff-a7a4db27e720b5572001d15719153b54), but without any comment as to why.

Reply:
I think it makes sense to move the validity checks below above `codegen_fn_attrs` and early-return if any of them fail. Then the check for `target_feature` can fit into the same loop. We should also have checks for the other, missing, attributes, but that could happen in another PR.

Reply:
I've changed the PR to what I think you've suggested here.